libs/os/dir.cpp

```diff
@@ -0,0 +1,40 @@
+#include "dir.h"
+
+#include <QDir>
+#include <QString>
+#include <QStringList>
+#include <QFileInfo>
+
+Directory::Directory(const char* name) {
+    this->dir = new QDir(name);
+}
+
+bool Directory::good() {
+    if(this->dir) {
+        QDir* dir_instance = reinterpret_cast<QDir*>(this->dir);
+        return dir_instance->exists();
+    }
+    return false;
+}
+
+void Directory::close() {
+    if(this->dir) {
+        QDir* dir_instance = reinterpret_cast<QDir*>(this->dir);
+        delete dir_instance;
+        this->dir = nullptr;
+    }
+}
+
+const char* Directory::read_and_increment(){
+    if(this->dir) {
+        QDir* dir_instance = reinterpret_cast<QDir*>(this->dir);
+        QStringList entryList = dir_instance->entryList(QDir::Dirs | QDir::Files | QDir::NoDotAndDotDot);
```

Comment:
This is srs perf penalty. E.g. with 1k entries in a browsed folder level at least 1M allocations will occur.

Reply:
changed